src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs

```diff
@@ -129,5 +129,24 @@ internal static bool IsPrimitiveType(this CorElementType et)
         [Intrinsic]
         internal static bool IsKnownConstant(int t) => false;
 #pragma warning restore IDE0060
+
+        private static unsafe void MemSet(ref byte dest, byte value, nuint size)
+        {
+            if (size > 0)
+            {
+                _ = dest;
```

Comment:
What's the purpose of the read? Maintaining null ref behavior?

Reply:
Yep. `SpanHelpers.Fill` and `Buffer.Memmove` may not provide needed guarantees for nullchecks. At least, `Buffer.Memmove` definitely does not - it may just call the native memmove via pinvoke.

Also, both impl may decide to fill/copy memory backwards (not uncommon for memcpy/memset) which may break null detection by dereferencing the guard page

Reply:
Looks like Roslyn removes it as unused 😐 

![image](https://github.com/dotnet/runtime/assets/523221/766f5e54-d869-456b-929f-ee9d45f15f15)

Reply:
That makes sense for normal locals. Seems over aggressive / behavior changing for a ref.

Reply:
Filed https://github.com/dotnet/roslyn/issues/72165

Reply:
Can work around with a helper method that just reads and returns it with `MI[AI]`, runtime can surely optimise it, and it allows C# to be happy.
```csharp
static byte Read(ref readonly b) => b;

...
_ = Read(in dest); //and similar
```

Reply:
> Filed https://github.com/dotnet/roslyn/issues/72165

I do not expect it to be fixed. C# language model assumes that `ref`s are never null and reading `ref`s is side-effect free. We had a discussion along these lines number of times.